src/lib.rs

```diff
@@ -1,49 +1,164 @@
+
+        // Debug
+        rsrc.iter().for_each(
+            |kv| tracing::debug!(target: "otel::setup::resource", key = %kv.0, value = %kv.1),
+        );
+
+        rsrc
+    }
+}
+
+#[derive(Debug)]
+pub struct ServiceInfoDetector {
+    fallback_service_name: &'static str,
+    fallback_service_version: &'static str,
+}
+
+impl ResourceDetector for ServiceInfoDetector {
+    fn detect(&self, _timeout: std::time::Duration) -> Resource {
+        let service_name = std::env::var("OTEL_SERVICE_NAME")
+            .or_else(|_| std::env::var("SERVICE_NAME"))
+            .or_else(|_| std::env::var("APP_NAME"))
+            .ok()
+            .or_else(|| Some(self.fallback_service_name.to_string()))
+            .map(|v| {
+                opentelemetry_semantic_conventions::resource::SERVICE_NAME.string(v)
+            });
+        let service_version = std::env::var("OTEL_SERVICE_VERSION")
+            .or_else(|_| std::env::var("SERVICE_VERSION"))
+            .or_else(|_| std::env::var("APP_VERSION"))
+            .ok()
+            .or_else(|| Some(self.fallback_service_version.to_string()))
+            .map(|v| {
+                opentelemetry_semantic_conventions::resource::SERVICE_VERSION.string(v)
+            });
+        Resource::new(vec![service_name, service_version].into_iter().flatten())
+    }
+}
+
+pub fn build_logger_text<S>(
+    log_level: tracing::Level,
+) -> Box<dyn Layer<S> + Send + Sync + 'static>
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+{
+    if cfg!(debug_assertions) {
+        Box::new(
+            tracing_subscriber::fmt::layer()
+                .pretty()
+                .with_line_number(true)
+                .with_thread_names(true)
+                .with_span_events(FmtSpan::NEW | FmtSpan::CLOSE)
+                .with_timer(tracing_subscriber::fmt::time::uptime()),
+        )
+    } else {
+        Box::new(
+            tracing_subscriber::fmt::layer()
+                .json()
+                .with_timer(tracing_subscriber::fmt::time::uptime())
+                .with_span_events(FmtSpan::NEW | FmtSpan::CLOSE)
+                .with_writer(std::io::stdout.with_max_level(log_level)),
+        )
+    }
+}
+
 pub fn init_tracing_with_fallbacks(
     log_level: tracing::Level,
     fallback_service_name: &'static str,
     fallback_service_version: &'static str,
 ) {
-    let otel_rsrc = DetectResource::default()
-        .with_fallback_service_name(fallback_service_name)
-        .with_fallback_service_version(fallback_service_version)
-        .build();
+    let otel_rsrc =
+        DetectResource::new(fallback_service_name, fallback_service_version).build();
     let otel_tracer =
         otlp::init_tracer(otel_rsrc, otlp::identity).expect("setup of Tracer");
-    let otel_layer = tracing_opentelemetry::layer().with_tracer(otel_tracer);
-
-    opentelemetry::global::set_text_map_propagator(
-        propagation::TextMapSplitPropagator::default(),
-    );
```

Comment:
The point of this `TextMapSplitPropagator` was to support multiple context extractors while only using one context injector.

I think it may still be relevant for Viaplay.

Reply:
```rust
// Configure the global propagator based on content of the env variable [OTEL_PROPAGATORS](https://opentelemetry.io/docs/concepts/sdk-configuration/general-sdk-configuration/#otel_propagators)
/// Specifies Propagators to be used in a comma-separated list.
/// Default value: `"tracecontext,baggage"`
/// Example: `export OTEL_PROPAGATORS="b3"`
/// Accepted values for `OTEL_PROPAGATORS` are:
///
/// - "tracecontext": W3C Trace Context
/// - "baggage": W3C Baggage
/// - "b3": B3 Single (require feature "zipkin")
/// - "b3multi": B3 Multi (require feature "zipkin")
/// - "xray": AWS X-Ray (require feature "xray")
/// - "ottrace": OT Trace (third party) (not supported)
/// - "none": No automatically configured propagator.
///
/// # Errors
///
/// Will return `TraceError` if issue in reading or instanciate propagator.
pub fn init_propagator() -> Result<(), TraceError> {
    let value_from_env = std::env::var("OTEL_PROPAGATORS")
        .unwrap_or_else(|_| "tracecontext,baggage".to_string());
    let propagators: Vec<(Box<dyn TextMapPropagator + Send + Sync>, String)> =
        value_from_env
            .split(',')
            .map(|s| {
                let name = s.trim().to_lowercase();
                propagator_from_string(&name).map(|o| o.map(|b| (b, name)))
            })
            .collect::<Result<Vec<_>, _>>()?
            .into_iter()
            .flatten()
            .collect();
    if !propagators.is_empty() {
        let (propagators_impl, propagators_name): (Vec<_>, Vec<_>) =
            propagators.into_iter().unzip();
        tracing::debug!(target: "otel::setup", OTEL_PROPAGATORS = propagators_name.join(","));
        let composite_propagator = TextMapCompositePropagator::new(propagators_impl);
        opentelemetry::global::set_text_map_propagator(composite_propagator);
    }
    Ok(())
}
```

Reply:
Personally I would rather keep `TextMapSplitPropagator` and drop current `init_propagator`

Reply:
`init_propagator` uses `TextMapCompositePropagator` which is different from our custom `TextMapSplitPropagator`.

The main difference is that `TextMapCompositePropagator` uses the same extractors and injectors.

We could have a chat if you want to understand the details better.

Reply:
> from our custom TextMapSplitPropagator.

Isn't ours right, is from otel package. TextMapCompositePropagator chains multiple [TextMapPropagator](https://vivint-smarthome.github.io/opentelemetry-stackdriver/opentelemetry/api/propagation/text_map_propagator/trait.TextMapPropagator.html) propagators together, injecting or extracting by their respective HTTP header names. I don't understand in which case `TextMapCompositePropagator` would be less useful than `TextMapSplitPropagator` for generic cases.

Reply:
`TextMapSplitPropagator` I implemented myself

Reply:
`TextMapCompositePropagator` uses the same set of propogators to handle both extraction and injection.

`TextMapSplitPropagator` uses different extractors and injectors.

We need to support at least three formats when doing extraction: `otlp`, `b3` and `b3-multi`. But when doing injection we only need `otlp` headers.

So, to support this model I wrote `TextMapSplitPropagator` using `TextMapCompositePropagator` as an example.

Reply:
`TextMapCompositePropagator` could still work for us, but:
 * we'll have to make sure that `OTEL_PROPAGATORS` contains all relevant propogators;
 * and we'll inject useless `b3` and `b3-multi` headers to all outgoing requests and to the response.